Proposal: Configmaps populating environment variables

fraenkel opened the pull request:
Proposal for configmap as environment variables

for #26299

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37295)
<!-- Reviewable:end -->

fraenkel committed:
Proposal: Configmaps populating environment variables

dims commented:
@k8s-bot ok to test

thockin commented on docs/proposals/envvar-configmap.md:
```diff
@@ -0,0 +1,162 @@
+## Proposed Design
+
+Containers can specify the ConfigMaps that are consumed as environment variables.
+Each key defined in the ConfigMap's `Data` object must be a "C" identifier.
```
or else ?  What happens if the data key is not a C ident?

thockin commented on docs/proposals/envvar-configmap.md:
```diff
@@ -0,0 +1,162 @@
+// EnvFromSource represents the source of a set of ConfigMaps
+type EnvFromSource struct {
+  // A string to place in front of every key
+  Prefix string
```
YAGNI ?  what's the motivation for this?

fraenkel replied:
See the last section of the proposed design. The prefix is to prevent collision of keys amongst multiple configmap and to make it obvious to the user where a given environment variable was derived from.

pmorie replied:
@thockin If we allow you to inject multiple configmaps / secrets and later services in this manner, there needs to be a way to disambiguate each block of environment variables to avoid collisions.

smarterclayton replied:
E.g.

MYSQL_ and URL
REDIS_ and URL

Asssuming people don't mingle config maps arbitrarily

thockin commented on docs/proposals/envvar-configmap.md:
```diff
@@ -0,0 +1,162 @@
+
+type Container struct {
+  // List of sources to populate environment variables in the container.
+  // The keys defined within a source must be a C_IDENTIFIER.
```
Spec what happens if it is not a C_IDENTIFIER.  Is it dropped? Fails to start the pod?  Modified in some way?

thockin commented on docs/proposals/envvar-configmap.md:
```diff
@@ -0,0 +1,162 @@
+metadata:
+  name: etcd-env-config
+data:
+  number-of-members: "1"
```
almost none of these keys are C identifiers...

thockin commented on docs/proposals/envvar-configmap.md:
```diff
@@ -0,0 +1,162 @@
+    - containerPort: 2380
+      protocol: TCP
+    envFrom:
+    - prefix: cm1.
```
periods are also not C_IDENTIFIERS

thockin commented:
That's good in theory, but do we actually need it?

On Tue, Nov 22, 2016 at 9:38 AM, Michael Fraenkel <notifications@github.com>
wrote:

> *@fraenkel* commented on this pull request.
> ------------------------------
>
> In docs/proposals/envvar-configmap.md
> <https://github.com/kubernetes/kubernetes/pull/37295>:
>
> > +
> +To prevent collisions amongst multiple ConfigMaps, each defined ConfigMap can have an associated prefix that is appended to each key in that ConfigMap. This also makes it obvious to the user where a given environment variable definition came from.
> +
> +### Kubectl updates
> +
> +The describe command will display the configmap name that have been defined as part of the environment variable section.
> +
> +### API Resource
> +
> +A new `EnvFromSource` type containing a `ConfigMapRef` will be added to the `Container` struct.
> +
> +```go
> +// EnvFromSource represents the source of a set of ConfigMaps
> +type EnvFromSource struct {
> +  // A string to place in front of every key
> +  Prefix string
>
> See the last section of the proposed design. The prefix is to prevent
> collision of keys amongst multiple configmap and to make it obvious to the
> user where a given environment variable was derived from.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/37295>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVALshPNpJ9vSb7POAE-_Nz1JULNsks5rAyiCgaJpZM4K5pBJ>
> .
>

fraenkel commented:
As far as the prefixes are concerned @pmorie asked for the ability to track where values came from.

We can either 
1. remove prefixes all together
1. make them optional
1. make them required

For the 2nd or 3rd case above, prefixes can also be required to be unique.

This decision is whether there is value it guaranteeing no collisions and identifying where they came from.

fraenkel committed:
Updated based on thockin's comments

Specify what happens when an invalid key is present
Use valid "C" identifiers in the example

dhoer commented:
I'm not sure this is the appropriate place to talk about this.  If not, please point me in the right direction.

First, I want to thank you for tackling this issue.  This is a much needed enhancement.   But I do have some concerns about this approach.  

When you say c_identifier, does that mean keys can be uppercase with underscores?  I think the implementation shouldn't care how your identifiers are formatted. Today, root data can't support [env variable names](https://google.github.io/styleguide/shell.xml?showone=Constants_and_Environment_Variable_Names#Constants_and_Environment_Variable_Names):

```
$ kubectl create configmap special-config --from-literal=SPECIAL_HOW=very
The ConfigMap "special-config" is invalid: data[SPECIAL_HOW]: Invalid value: "SPECIAL_HOW": must have at most 253 characters and match regex \.?[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*
```

But configmap files can be anything:
```
$ echo "# a comment" > special-config.properties
$ echo "" >> special-config.properties
$ echo "SPECIAL_HOW=very" >> special-config.properties
$ echo "Non-Conforming_KEY=I'm special" >> special-config.properties
$ kubectl create configmap special-config --from-file=special-config.properties
configmap "special-config" created
$ kubectl get configmaps special-config -o yaml
apiVersion: v1
data:
  special-config.properties: |
    # a comment

    SPECIAL_HOW=very
    Non-Conforming_KEY=I'm different
kind: ConfigMap
metadata:
  creationTimestamp: 2016-11-22T17:33:21Z
  name: special-config
  namespace: my-app
  resourceVersion: "9729470"
  selfLink: /api/v1/namespaces/my-app/configmaps/special-config
  uid: c332e35a-b0d9-11e6-8c60-065eece225bf
```
It would be nice to process the env file ignoring comments and blank lines as well as not caring about the key value naming convention used. 

Handling of values with environment variables referenced would be a bonus as well:
ENV_KEY=my_${SPECIAL_HOW}_special_value

So now you can reference the configmap object:
```
  - envFrom:
      configMap: special-config/special-config.properties
```

Then some report (debug?) can show the rendered resulting maps somehow:
```
  - configMap: 
     - special-config/special-config.properties: 
        SPECIAL_HOW=very
        Non-Conforming_KEY=I'm different
  env:
    SPACE_SHIPS=1
    SHIP_TYPE=battle-cruiser
    WEAPON=laser-cannon
    SPECIAL_HOW=not very
```
And finally, the merge map (where repeated subsequent key values are overwritten):
```
env:
    Non-Conforming_KEY=I'm different
    SPECIAL_HOW=not very
    SPACE_SHIPS=1
    SHIP_TYPE=battle-cruiser
    WEAPON=laser-cannon
```
# Key advantages of this approach
- separation of concern for what naming convention customers use - no longer need to care
- no need to format keys to match k8s requirements
- allow for comments and blanks lines in environment files which are naturally expected 
- allow one configmap bucket for multiple env files that pertain to deployment (let's say one property file was for putting the app in debug mode, it would be nice to have that in the same configmap as the core app and toggle it on and off in the deployment file)

thockin commented:
On Tue, Nov 22, 2016 at 10:20 AM, Dennis Hoer <notifications@github.com> wrote:
>
> I'm not sure this is the appropriate place to talk about this. If not, please point me in the right direction.
>
> First, I want to thank you for tackling this issue. This is a much needed enhancement. But I do have some concerns about this approach.
>
> When you say c_identifier, does that mean keys can be uppercase with underscores? I think the implementation shouldn't care how your identifiers are formatted. Today, root data can't support env variable names:
>
> $ kubectl create configmap special-config --from-literal=SPECIAL_HOW=very
> The ConfigMap "special-config" is invalid: data[SPECIAL_HOW]: Invalid value: "SPECIAL_HOW": must have at most 253 characters and match regex \.?[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*

ConfigMap keys were expanded to support uppercase and underscores a
release or two ago - what release are you on?

> But configmap files can be anything:
>
> $ echo "# a comment" > special-config.properties
> $ echo "" >> special-config.properties
> $ echo "SPECIAL_HOW=very" >> special-config.properties
> $ echo "Non-Conforming_KEY=I'm special" >> special-config.properties
> $ kubectl create configmap special-config --from-file=special-config.properties
> configmap "special-config" created
> $ kubectl get configmaps special-config -o yaml
> apiVersion: v1
> data:
>   special-config.properties: |
>     # a comment
>
>     SPECIAL_HOW=very
>     Non-Conforming_KEY=I'm different

This is a different design -I don't like the extra level of
indirection, but I've done worse.  Looking for other opinions...

> kind: ConfigMap
> metadata:
>   creationTimestamp: 2016-11-22T17:33:21Z
>   name: special-config
>   namespace: my-app
>   resourceVersion: "9729470"
>   selfLink: /api/v1/namespaces/my-app/configmaps/special-config
>   uid: c332e35a-b0d9-11e6-8c60-065eece225bf
>
> It would be nice to process the env file ignoring comments and blank lines as well as not caring about the key value naming convention used.
>
> Handling of values with environment variables referenced would be a bonus as well:
> ENV_KEY=my_${SPECIAL_HOW}_special_value

This is sort of hard to reason about, IMO - I'd like to see good justification.

> So now you can reference the configmap object:
>
>   - envFrom:
>       configMap: special-config/special-config.properties
>
> Then some report (debug?) can show the rendered resulting maps somehow:
>
>   - configMap:
>      - special-config/special-config.properties:
>         SPECIAL_HOW=very
>         Non-Conforming_KEY=I'm different
>   env:
>     SPACE_SHIPS=1
>     SHIP_TYPE=battle-cruiser
>     WEAPON=laser-cannon
>     SPECIAL_HOW=not very
>
> And finally, the merge map (where repeated subsequent key values are overwritten):
>
> env:
>     Non-Conforming_KEY=I'm different
>     SPECIAL_HOW=not very
>     SPACE_SHIPS=1
>     SHIP_TYPE=battle-cruiser
>     WEAPON=laser-cannon
>
> Key advantages of this approach
>
> separation of concern for what naming convention customers use - no longer need to care

Well we DO need to care - the key has to end up being a valid
environment variable, which is currently specced as C_IDENTIFIER.
Anything that is not conforming to that won't work.

> no need to format keys to match k8s requirements

see previous :)

> allow for comments and blanks lines in environment files which are naturally expected

YAML allows comments

> allow one configmap bucket for multiple env files that pertain to deployment (let's say one property file was for putting the app in debug mode, it would be nice to have that in the same configmap as the core app and toggle it on and off in the deployment file)

The main advantage I see of reading a key from configmap is that a
rollout can keep the same configmap and change which key is being
used, same as config blobs.  The downside is it is another inderection
and harder to reason about.

dhoer commented:
>  ConfigMap keys were expanded to support uppercase and underscores a release or two ago - what release are you on?

I thought I was on 1.4, but it looks like the cluster 1.3.  Good to know...

> This is a different design -I don't like the extra level of indirection, but I've done worse.  Looking for other opinions...

I just don't like the fact that adding a customer's env file requires a pre-processing step to read in each key value pair.  Whereas the --from-file clause would just require pointing to it, no further processing required.

> This is sort of hard to reason about, IMO - I'd like to see good justification.

I think templates could handle this...

> Well we DO need to care - the key has to end up being a valid environment variable, which is currently specced as C_IDENTIFIER. Anything that is not conforming to that won't work.

Related to me being frustrated by 1.3 constraints.  As long as C_IDENTIFIER doesn't constrain what posix allows for env names.  

>  YAML allows comments

Not sure how create configmap could add that:
```
$ kubectl create configmap special-config --from-literal="# a comment"
error: invalid literal source # a comment, expected key=value
```

> The main advantage I see of reading a key from configmap is that a rollout can keep the same configmap and change which key is being used, same as config blobs.  

That is what I like about it.  I can take there files as is with no pre-processing...

> The downside is it is another indirection and harder to reason about.

I think it would be nice to treat configmap as a bucket of environment configs to be included in deployment without having to update the configmap.

fraenkel commented:
Do we need input from @pmorie to decide on the prefix requirement?
Just trying to understand how we can move this forward since I believe it is the only remaining issue.

dhoer commented:
@thockin I just can't let this go...

> The main advantage I see of reading a key from configmap is that a rollout can keep the same configmap and change which key is being used, same as config blobs.  The downside is it is another inderection and harder to reason about.

There is a whole cottage industry built around how to convert env files into kubernetes configmaps and secrets.  That is a major code smell.  IMHO there should be no pre-processing required to add them. Now here we are talking about having the cottage industry refactor to handle prefixes? 

Please look at the big picture and imagine bringing on thousands of projects to k8s.  Does the current requirement for pre-processing using cottage industry or homegrown solutions still make sense because of concern of another layer of indirection?

thockin commented:
I'm not sure I understand.  Whether you do one value per key or one key
with many values, the tooling can be adapted to make it simple.

On Wed, Nov 23, 2016 at 10:06 AM, Dennis Hoer <notifications@github.com>
wrote:

> @thockin <https://github.com/thockin> I just can't let this go...
>
> The main advantage I see of reading a key from configmap is that a rollout
> can keep the same configmap and change which key is being used, same as
> config blobs. The downside is it is another inderection and harder to
> reason about.
>
> There is a whole cottage industry built around how to convert env files
> into kubernetes configmaps and secrets. That is a major code smell. IMHO
> there should be no pre-processing required to add them. Now here we are
> talking about having the cottage industry refactor to handle prefixes?
>
> Please look at the big picture and imagine bringing on thousands of
> projects to k8s. Does the current requirement for pre-processing using
> cottage industry or homegrown solutions still make sense because of concern
> of another layer of indirection?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/37295#issuecomment-262589589>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVOyYkBgvmBFqAB9WwAM6ZelJGALiks5rBICUgaJpZM4K5pBJ>
> .
>

pmorie commented:
I would strongly prefer to have a way to qualify the environment variables created in this manner via a prefix or format string.  It is entirely possible for 2 resources that are injected in this manner (two configmaps, two secrets, secret and a configmap, etc) to have the same keys, which will result in user not getting what is expected, given the information in the pod spec.  I would accept this prefix or format string being optional, if it is not desired, but I think many people would prefer to use it in cases where multiple configmaps / secrets are injected.

@dhoer I don't understand your comments -- can you elaborate?

dhoer commented:
@pmorie I will try.  I wanting k8s to be able read in an env file like this:

Create configmap:
`kubectl create configmap game-config --from-file=./env/game.env`

Reference configmap:
```yaml
- envFrom:
      configMap: game-config/game.env
```

Comments and blank spaces would be handled automatically, just like in docker env-file:
https://docs.docker.com/compose/env-file/.

For consistency, it would be nice for secrets to have the same behavior.

Create secret:
`kubectl create secret game-config --from-file=./secret/too-many-secrets.env`

Reference secret:
```yaml
- envFrom:
      secret: game-config/too-many-secrets.env
```

This eliminates the required preprocessing step to convert name value pairs into configmap or secret, e.g., https://github.com/dhoer/k8s-env-gen.

pmorie commented:
@dhoer I think there is already an issue for that -- IMO that is orthogonal from this feature.  I'll attempt to find the issue for you.

fraenkel committed:
Allow prefixes to be optional

fraenkel committed:
Prefix are not unique, but must be a C identifier.

thockin commented:
@pmorie:
> I would strongly prefer to have a way to qualify the environment variables created in this manner via a prefix

Without out-of-bounds characters, you simply can not *guarantee* that there are no collisions.

if set A has FOO=BAR and set B has FOO=BAR : collision

if set A has FOO=BAR, prefix A_ and set B has A_FOO=BAR : collision

if set A has FOO=BAR, prefix A_ and set B has _FOO=BAR, prefix A : collision (and mandating non-empty prefixes is user-hostile anyway).

You can reduce the chances of collision but you still have to define what happens when they collide, and plumb the errors back to the user.  Given that, do prefixes REALLY matter?  Is there a use case to import the same configmap in two different pods with different prefixes?  It just feels like YAGNI.

thockin commented:
The real open question is whether teh configmaps look like this:

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: etcd-env-config
data:
  number_of_members: "1"
  initial_cluster_state: new
  initial_cluster_token: DUMMY_ETCD_INITIAL_CLUSTER_TOKEN
```

or

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: etcd-env-config
data:
  my-config: |
    number_of_members=1
    initial_cluster_state=new
    initial_cluster_token=DUMMY_ETCD_INITIAL_CLUSTER_TOKEN
```

Yes?

dhoer commented:
@thockin why not support both ways?
```yaml
envFrom:
  - configMapRef:
    name: etcd-env-config
```
or
```yaml
envFrom:
  - configMapRef:
    name: etcd-env-config
    key: my-config
```

I agree about YAGNI statement. I would probably have to refactor the Dockerfile to use prefix.  So I might as well create new key names anyway.  Thus, no need for prefix.

thockin commented:
"both" is the lazy answer with the highest bill to pay.

On Mon, Nov 28, 2016 at 2:25 PM, Dennis Hoer <notifications@github.com>
wrote:

> @thockin <https://github.com/thockin> why not support both ways?
>
> envFrom:
>   - configMapRef:
>     name: etcd-env-config
>
> or
>
> envFrom:
>   - configMapRef:
>     name: etcd-env-config
>     key: my-config
>
> I agree about YAGNI statement. I would probably have to refactor the
> Dockerfile to use prefix. So I might as well create new key names anyway.
> Thus, no need for prefix.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/37295#issuecomment-263414333>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVEZAN0uFFwsGLUKzw3PVK9Looa_Tks5rC1T3gaJpZM4K5pBJ>
> .
>

pmorie commented:
The use-case for the prefix is that you have two configmaps or two secrets with the **exact same** keys that you want to use.  Here's an example: you have two database services that have the exact same information stored in two different secrets, with the exact same set of keys in each secret, and you want to consume them both in this manner.  To do this, you have to have a prefix.  Make sense?

thockin commented:
Is this going to be common enough to justify the feature?  I'm not close
enough to the ServiceBroker work to know.

On Mon, Nov 28, 2016 at 3:11 PM, Paul Morie <notifications@github.com>
wrote:

> The use-case for the prefix is that you have two configmaps or two secrets
> with the *exact same* keys that you want to use. Here's an example: you
> have two database services that have the exact same information stored in
> two different secrets, with the exact same set of keys in each secret, and
> you want to consume them both in this manner. To do this, you have to have
> a prefix. Make sense?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/37295#issuecomment-263424800>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVLqypLZbNngYDJ7RqkjJ9LjgpqYtks5rC1-ygaJpZM4K5pBJ>
> .
>

dhoer commented:
@pmorie I'm not sure that makes sense to me.  Using 2 different configmaps or using --from-file (if supported) would solve this issue without the need to add prefix. I'm assuming each service runs in a container and each container's environment key values are isolated from each other.

fraenkel commented:
@dhoer You are looking at the wrong side of the services.  Think about an app using 2 databases. The config provided for each database may be in a configmap.

dhoer commented:
@thockin well you know I'd prefer the latter...

dhoer commented:
@thockin The latter would make it easy for kompose to map all env_file entries to a configmap.

thockin commented:
devil's advocate: it's pretty trivial to parse an env file and split it
into distinct keys, no?

On Mon, Nov 28, 2016 at 4:23 PM, Dennis Hoer <notifications@github.com>
wrote:

> @thockin <https://github.com/thockin> The latter would make it easy for
> kompose to map all env_file entries to a configmap.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/37295#issuecomment-263438226>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVMyYnAdUUk9IDaUCyvdoZ0D_p1zGks5rC3BlgaJpZM4K5pBJ>
> .
>

pmorie commented:
@thockin 

> Is this going to be common enough to justify the feature? I'm not close enough to the ServiceBroker work to know.

We found it to be really common in previous versions of OpenShift that people would want to consume a multiple instances of the same thing and we didn't have a good way to disambiguate the equivalent of this feature.  We had to do very ugly things to work around it.  I am fine going into terrible, eye-scorching detail if it is desired, but won't for now because the explanation requires some description that I doubt anyone will be super interested in reading if I can't make the point based on the merits alone :)

@dhoer 

> @pmorie I'm not sure that makes sense to me. Using 2 different configmaps or using --from-file (if supported) would solve this issue without the need to add prefix. I'm assuming each service runs in a container and each container's environment key values are isolated from each other.

The use cases I am talking about are ones where there are two different resources that you do not have the option to combine.  Where you as a user have discretion you can do whatever you want and make it so you don't need a prefix.  I think prefix should be optional, so if you don't want to use it, you shouldn't have to.  The use-cases I am interested in supporting correctly are ones where:

1.  You want to consume configmaps/secrets that you may not know the contents of in advance and have no control over in any event
2.  You may want to consume N number of configmaps/secrets that not only do you not know the contents of in advance, but might have the exact same set of keys that can collide

Let me say more about my specific motivation in this area.  I am working on the kubernetes service catalog; one of the value-adds of the catalog is that it will allow a user to make a binding to something represented by a kubernetes Service and be injected via secrets and configmaps with the information necessary to use that thing.  As I alluded to earlier in this comment, I have some history with problems like this from working on the first two versions of OpenShift.  It is very, very common for people to want to use multiple different but similar things or different instances of the same thing.  Examples:

1.  I want to use multiple different instances of an RDBMS and need the same type of information to connect to each of them
2.  I want to use multiple instances of `<your favorite nosql datastore here>` and need the same type of information to connect to each of them
3.  I want to use multiple MQ brokers and need the same type of information to connect to each of them

If I had to bet on the most common one of the above, it would be (3) with a bullet.  You find this pattern all over enterprise systems that have a microservicey architecture that communicates via MQ messages where each team / division runs their own set of brokers.  A close second is probably (1).

In these cases, the user typically does not have control at a level where they can combine these different sources of information into a single configmap or secret, and it is usually not desirable to do so.   That means that we need to have a good way to disambiguate things on the consumption end of concerns, which implies something like a prefix or format string.  I think an optional prefix is a happy medium.  If you don't want it, don't use it.  You can mix using it and not using it, if you know what you're doing.  If you decide to use it, you will probably be glad that it's there.

Does that clarify things at all?

pmorie commented:
cc @jpeeler @jwforres @bparees

dhoer commented:
@pmorie I suppose. I never thought of handling those scenarios this way...

bparees commented:
Can someone (or ideally the proposal) articulate why the existing valueFrom mechanism is considered insufficient for the intended use case?  Is it a desire to more easily reference a chunk of config?  to limit the customization needed in the container definition?

the valueFrom mechanism at a minimum solves the prefixing problem since the key name becomes arbitrary.

```
containers:
    - name: test-container
      image: gcr.io/google_containers/busybox
      command: [ "/bin/sh", "-c", "env" ]
      env:
        - name: SPECIAL_LEVEL_KEY
          valueFrom:
            configMapKeyRef:
              name: special-config
              key: special.how
        - name: SPECIAL_TYPE_KEY
          valueFrom:
            configMapKeyRef:
              name: special-config
              key: special.type
```

bparees commented on docs/proposals/envvar-configmap.md:
```diff
@@ -0,0 +1,162 @@
+Containers can specify the ConfigMaps that are consumed as environment variables.
+Each key defined in the ConfigMap's `Data` object must be a "C" identifier. If an invalid key is present, the container will fail to start.
+
+Environment variables are currently defined by servies or the `Env` object in a container, where the `Env` takes precedence.
```
s/servies/services/

fraenkel committed:
Spelling correction

pmorie commented:
@bparees Many people consider the existing mechanism to be burdensome to use.  I view the proposed mechanism as a shortcut around having to repeat yourself if all you want is to dump keys into env; if you need finer-grained control you probably want the current mechanism.

thockin commented:
> Does that clarify things at all?

@pmorie FTW.  I'm sold.

pmorie commented:
@kubernetes/api-review-team PTAL

LGTM personally.

pmorie commented:
@fraenkel Actually, upon further consideration, the motivation I described should probably make it inot the propoxal in some form.

thockin commented:
ping me when ready for a another read

On Tue, Nov 29, 2016 at 12:17 PM, Paul Morie <notifications@github.com>
wrote:

> @fraenkel <https://github.com/fraenkel> Actually, upon further
> consideration, the motivation I described should probably make it inot the
> propoxal in some form.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/37295#issuecomment-263686077>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVOc98E17fPtbJ91Px1M-6BhHihqOks5rDIhugaJpZM4K5pBJ>
> .
>

fraenkel committed:
Add Design Point section

fraenkel commented:
@thockin @pmorie PTAL.  I added a design point section to add the additional detail.

pmorie commented on docs/proposals/envvar-configmap.md:
```diff
@@ -0,0 +1,168 @@
+  // A string to place in front of every key. Must be a C_IDENTIFIER.
+  // +optional
+  Prefix string `json:"prefix,omitempty"`
+	// The ConfigMap to select from
```
Nit: some spacing problems in this go section.

pmorie commented on docs/proposals/envvar-configmap.md:
```diff
@@ -0,0 +1,168 @@
+  // +optional
+  EnvFrom []EnvFromSource `json:"envFrom,omitempty"`
+}
+
```
dead line

pmorie commented on docs/proposals/envvar-configmap.md:
```diff
@@ -0,0 +1,168 @@
+	ConfigMapRef *LocalObjectReference `json:"configMapRef,omitempty"`
+}
+
+
```
dead line

pmorie commented on docs/proposals/envvar-configmap.md:
````diff
@@ -0,0 +1,168 @@
+}
+
+```
+
````
I usually try to describe the Kubelet changes at a very high level in a proposal.

pmorie reviewed (commented):
Would you wrap lines at 80 chars?

fraenkel committed:
Prettify the markdown

fraenkel commented:
@pmorie Changes have been made. The detail provided was to cover all the concerns raised in the original PR.

bgrant0607 commented:
cc @pwittrock

bgrant0607 commented:
I suggest converging on the proposal here, then close the PR and copy the final result to the community repo.

erictune commented on docs/proposals/envvar-configmap.md:
```diff
@@ -0,0 +1,182 @@
+
+### Kubectl updates
+
+The describe command will display the configmap name that have been defined as
```
name does not uniquely identify which configmap was used.  You would need resource version and UID to confirm a match between the pod and a configmap object.

erictune replied:
I'm not objecting to this proposal: AIUI, it is the same as what we already support but less verbose.  I'm just saying that this describe behavior might not be that useful.

fraenkel replied:
It is identical to ConfigMapKeySelector. Once LocalObjectReference fixes it's TODO, then we get more data. But given we only have 1 version and its already scoped to the namespace of the Pod it is unique enough.

Adding support to describe is just filling out the rest of what can be presented to the user or else the user might get confused when they don't see the ConfigMaps tied to their Pods. It can be removed, just not sure how a user would feel if they are missing information behind the environment variables that magically appear.

dhoer commented:
@bparees To answer your question; yes, it is to limit the migration steps required for env files.  I'm not sure about CoreOS, but with Docker and Docker Compose env files seem natural and easy to use.  With K8s, it doesn't.  It requires a process to map the env files into deployment env section like what [kompose](https://github.com/kubernetes-incubator/kompose) does. But kompose converts both env vars and env files into env entries losing context if the env was from an env variable or and env file.

One way to handle env files that keeps the relationship is to use configmap.  Docker env vars can be mapped to env section, and env files can be mapped to configmap with envFrom in deployment yaml.  While you can currently map env values to configmap like you show, it is very verbose. A configmap with 100 env value pairs would equal 500 lines in deployment yml.  Doing it the kompose way would be 200 lines.  Using envFrom, 3 lines.

thockin commented:
@dhoer: if there were a `kubectl create configmap` mode that read a file an env file and created the equivalent configmap (1 key per key), would you be satisfied?

thockin commented:
I have no further comments.  Please move to community repo.

dhoer commented:
@thockin Yes, that would be a workable solution.

thockin commented:
Great - please include that in this proposal when it moves to new repo.

On Tue, Dec 6, 2016 at 7:56 AM, Dennis Hoer <notifications@github.com>
wrote:

> @thockin <https://github.com/thockin> Yes, that would be a workable
> solution.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/37295#issuecomment-265187490>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVFDb6hG_i3mA7Z_tQfPiFtCjmhEAks5rFYWwgaJpZM4K5pBJ>
> .
>

fraenkel commented:
The proposal has moved and I would like to continue this discussion over there.

troyharvey commented:
[tldr](https://gist.github.com/troyharvey/4506472732157221e04c6b15e3b3f094)
